README.md
# Karoo Tools

Karoo Tools help you prepare your data for Machine Learning runs. Each Tool has one function. The intent is easy to
use, simple to learn. Let me know if I have achieved this goal. An introduction to each Tool is included in the header. 

All data is anticipated to be in the following fomat:
 - comma separated values .csv
 - Apple's Numbers spreadsheet is *not* recommended as their line breaks are non-ASCII standard
 - header contains alpha-numeric names of features (variables)
 - right-most column is the solution (label)
 - left-most column may be the ID (used in the pipeline)
 - rows are instances, but not necessarily steps in time


To learn how to use each tool:

	less karoo_[tool].py

If you combine the fundamental tools Clean, Sort, and Normalize, you'll see the progression of names:

	sample_data.csv
	sample_data-CLEAN.csv
	sample_data-CLEAN-SORT.csv
	sample_data-CLEAN-SORT-NORM.csv

Such that the final dataset is the one prepared for your Machine Learning algorithm.

More tools coming ...


RELEASE_NOTES.txt
2018 03/26b

Got the karoo_pipeline.py working with or without the left-most column being an ID.

A typo fixed in karoo_clean.py. Several minor, aesthetic updates to all the tools.


2018 03/26a

Spent the better part of today working throgh my archive of working but could-be-better tools, giving them the final
touches to bring them into the Karoo family. Now each has a proper old-school-is-cool banner, nearly identical code
flow, and for the most part, similar naming scheme for the variables. This creates some familiarity to the total
package such that if you dive into one, modifying it for your own local needs, it will be easy to do so again with
the next.

I have yet to upload the karoo_pipeline as I have one more modification to take it out of its LIGO foundation and
into a more generic functionality.


2018 03/25

No notes just yet, sorry :(

Kai Staats

www.kaistaats.com/research/
www.kaistaats.com/film/


karoo_pipeline.py
# Karoo GP Pipeline
# by Kai Staats; see LICENSE.md
# see LICENSE.md
# version 2018 03/26

import os
import csv
import sys
import numpy as np
import sympy as sp

np.set_printoptions(linewidth = 320) # set the terminal to print 320 characters before line-wrapping in order to view Trees

'''
This script is currently designed to work with binary classification only. It can be readily modified to work with
multiclass classifiction by adding upper and lower thresholds for each additional classification bin (below).

Once you have applied Karoo GP and are satisified with the quality of the evolved multivariate expression, you may be
ready to move that function into the real world. This script gives you a working foundation to do just that, by pasting
the Sympified (or raw) multivariate expression (e.g. a**2 + 2b * -3c) into the supplied interface. The script then 
executes the supplied function against the data, row by row, and talies the result into one of two bins (for the 
default binary classification). You can readily extract the ID numbers from each bin and determine how the each data
point has been classified.

	python karoo_pipeline.py [path/dataset.csv]
	
* no sample expression or data provided at this time
'''

### USER INTERACTION ###

if len(sys.argv) == 1: print '\n\t\033[31mERROR! You have not assigned an input file. Try again ...\033[0;0m'; sys.exit()
elif len(sys.argv) > 2: print '\n\t\033[31mERROR! You have assigned too many command line arguments. Try again ...\033[0;0m'; sys.exit()
else: filename = sys.argv[1]

os.system('clear')
		
print '\n\033[36m\033[1m'
print '\t **   **   ******    *****    ******    ******       ******   **'
print '\t **  **   **    **  **   **  **    **  **    **     **    **  **'
print '\t ** **    **    **  **   **  **    **  **    **     **    **  **'
print '\t ****     ********  ******   **    **  **    **     *******   **'
print '\t ** **    **    **  ** **    **    **  **    **     **        **'
print '\t **  **   **    **  **  **   **    **  **    **     **        **'
print '\t **   **  **    **  **   **  **    **  **    **     **        **'
print '\t **    ** **    **  **    **  ******    ******      **        ******'
print '\033[0;0m'
print '\t\033[36m Apply your evolved GP expression to a realworld pipeline - by Kai Staats\033[0;0m\n'

while True:
	try:
		algo_paste = raw_input('\t Enter the Sympified expression from Karoo GP: ')
		if algo_paste == '': raise ValueError()
		break
	except ValueError: print '\n\t\033[32m An empty expression is not going to get you very far. Try again ...\n\033[0;0m'
	except KeyboardInterrupt: sys.exit()

#while True:
#	try:
#		print_out = raw_input('\n\t Print the results to screen? y/n (default n): ')
#		if print_out not in ['y','n','']: raise ValueError()
#		print_out = print_out or 'n'; break
#	except ValueError: print '\033[32mSelect from the options given. Try again ...\n\033[0;0m'

precision = 4 # set the number of floating points


### LOAD DATA ###

data = np.loadtxt(filename, delimiter = ',', dtype = str) # load the data
header = data[0] # store the header for each column
data = np.delete(data,0,0) # delete the header from the data
data = data.astype(np.float) # convert the remaining array from string to float

data_rows = data.shape[0]
data_cols = data.shape[1]
data_dict = {} # prepare an empty dictionary
data_dict_array = np.array([]) # prepare an empty array to hold the dictionaries

for row in range(0, data_rows): # increment through each row of data
	for col in range(0, data_cols): # increment through each column
		data_dict.update( {header[col]:data[row,col]} ) # add a row of data into a dictionary	
		
	data_dict_array = np.append(data_dict_array, data_dict.copy()) # add a dictionary of data into the array

		
### CLASSIFY THE PIPELINE DATA ###

def fx_eval_subs(datum, algo_paste):

	'''
	Evaluation of the raw expression into a sympified expression was already done in Karoo GP. Therefore, we need only 
	assign each feature (column) with an associated data point (row) from the pipeline data, and then process.
	'''
	
	algo_sym = sp.sympify(algo_paste) # string converted to a functional expression
	subs = algo_sym.subs(datum) # process the expression against the datum
	if str(subs) == 'zoo': # result = 1 # TEST & DEBUG: print 'divide by zero', result; self.fx_karoo_pause(0)
		print '\n\t\033[31mERROR! Divide by zero\033[0;0m'; sys.exit()
		
	else: result = round(float(subs), precision) # force 'result' to the set number of floating points
		
	return result
	

# If using iPython, after the pipeline is executed, you can extract any given feature value from any given row:
#
#		data_dict_array[row]['key']
# 
# ... where 'key' is a column header (feature variable)

class_0 = []
class_1 = []

for row in range(0, data_rows):
	result = fx_eval_subs(data_dict_array[row], algo_paste) # process the expression against the test data
	
	if result <= 0: # test for class 0
		if header[0] == 'id': class_0.append([data_dict_array[row]['id']]) # record the ID if available
		else: class_0.append(row) # else record the .csv row number
		print '\t\033[36m data row', row, 'predicts class:\033[1m 0 as', result, '<=', 0, '\033[0;0m'
		
	elif result > 0: # test for class 1
		if header[0] == 'id': class_1.append([data_dict_array[row]['id']]) # record the ID if available		
		else: class_1.append(row) # else record the .csv row number
		print '\t\033[36m data row', row, 'predicts class:\033[1m 1 as', result, '>', 0, '\033[0;0m'
		
	else: print 'Whoa! The result is neither <= 0 nor > 0. The universe will implode in 5 ... 4 ... 3 ...'; sys.exit()
	

### SAVE THE OUTPUT ###

print '\n\t Writing', len(class_1), 'class 1 to disk'
target = open('class_1.csv', 'w'); target.close() # initialise class 1 .csv file
with open('class_1.csv', 'a') as csv_file:
	target = csv.writer(csv_file, delimiter=',')
	#target.writerows([['IDs for class 1 (label 1)']])
	target.writerows([class_1])

print '\n\t Writing', len(class_0), 'class 0 to disk'
target = open('class_0.csv', 'w'); target.close() # initialise class 0 .csv file
with open('class_0.csv', 'a') as csv_file:
	target = csv.writer(csv_file, delimiter=',')
	#target.writerows([['IDs for class 0 (label 0)']])
	target.writerows([class_0])




karoo_data_norm.py
# Karoo Data Normalisation
# by Kai Staats, MSc
# see LICENSE.md
# version 2018 03/26

import os
import sys
import numpy as np

np.set_printoptions(linewidth = 320) # set the terminal to print 320 characters before line-wrapping in order to view Trees

'''
This script works with a dataset to prepare a new, normalised dataset. It does so by comparing all values in each given 
column, finding the maximum and minimum values, and then modifying each value to fall between a high of 1 and low of 0. 
The modified values are written to a new file, the original remaining untouched. This script is often used *after* 
karoo_data_sort.py.

	python karoo_normalise.py sample.csv
	
The original dataset is left unaltered.
'''

### USER INTERACTION ###

if len(sys.argv) == 1: print '\n\t\033[31mERROR! You have not assigned an input file. Try again ...\033[0;0m'; sys.exit()
elif len(sys.argv) > 2: print '\n\t\033[31mERROR! You have assigned too many command line arguments. Try again ...\033[0;0m'; sys.exit()
else: filename = sys.argv[1]

os.system('clear')

print '\n\033[36m\033[1m'
print '\t **   **   ******    *****    ******    ******      **    **   ******    *****   **       **'
print '\t **  **   **    **  **   **  **    **  **    **     ***   **  **    **  **   **  ***     ***'
print '\t ** **    **    **  **   **  **    **  **    **     ****  **  **    **  **   **  ****   ****'
print '\t ****     ********  ******   **    **  **    **     ** ** **  **    **  ******   ** ** ** **'
print '\t ** **    **    **  ** **    **    **  **    **     **  ****  **    **  ** **    **  ***  **'
print '\t **  **   **    **  **  **   **    **  **    **     **   ***  **    **  **  **   **   *   **'
print '\t **   **  **    **  **   **  **    **  **    **     **    **  **    **  **   **  **       **'
print '\t **    ** **    **  **    **  ******    ******      **    **   ******   **    ** **       **'
print '\033[0;0m'
print '\t\033[36m Data Prep for Machine Learning in Python - by Kai Staats\033[0;0m\n'

menu = range(1,9)
while True:
	try:
		fp = raw_input('\t Enter number of floating points desired in normalised data (default 4): ')
		if fp not in str(menu) and fp not in '': raise ValueError()
		if fp == '0': fp = 1; break
		fp = fp or 4; fp = int(fp); break
	except ValueError: print '\n\t\033[32m Enter a number from 1 including 8. Try again ...\033[0;0m'
	

### LOAD THE DATA ###

print '\n\t\033[36m Loading dataset:', filename, '\033[0;0m'
data = np.loadtxt(filename, delimiter = ',', dtype = str) # load the data
header = data[0] # store the header for each column
data = np.delete(data,0,0) # delete the header from the data
data = data.astype(np.float) # convert the remaining array from string to float


### NORMALISE THE DATA ###

def normalise(array):

	'''
	The formula was derived from stn.spotfire.com/spotfire_client_help/norm/norm_normalizing_columns.htm 
	'''
	
	norm = []
	array_norm = []
	array_min = np.min(array)
	array_max = np.max(array)
	
	for col in range(1, len(array) + 1):
		# norm = float((array[col - 1] - array_min) / (array_max - array_min))
		norm = float(array[col - 1] - array_min) / float(array_max - array_min)
		norm = round(norm, fp) # force to 4 decimal points
		array_norm = np.append(array_norm, norm)
		
	return array_norm
	

data_out = np.zeros(shape = (data.shape[0], data.shape[1])) # build an array that matches the shape of the original

for col in range(data.shape[1] - 1): # count columns, but exclude the right-most labels column so as to not delete labels
	print '\t normalising column:', col
	
	colsum = []
	for row in range(data.shape[0]):
		colsum = np.append(colsum, data[row,col])
		
	data_out[:,col] = normalise(colsum) # add each normalised column of data
	
data_out[:,data.shape[1] - 1] = data[:,data.shape[1] - 1] # add the labels again


### SAVE THE MODIFIED DATA ###

data_out = np.vstack((header, data_out)) # re-attach the header to the data
file_tmp = filename.split('.')[0]
np.savetxt(file_tmp + '-NORM.csv', data_out, delimiter = ',', fmt='%s')

print '\n\t\033[36m The normlised dataset has been written to the file:', file_tmp + '-NORM.csv', '\033[0;0m'




karoo_data_sort.py
# Karoo Data Sort
# by Kai Staats, MSc and Arun Kumar, PhD
# see LICENSE.md
# version 2018 03/26

import os
import sys
import numpy as np

np.set_printoptions(linewidth = 320) # set the terminal to print 320 characters before line-wrapping in order to view Trees

'''
In machine learning, it is often the case that your engaged dataset is derived from a larger parent. In constructing 
a subset, if we grab a series of datapoints (rows in a .csv) from the larger dataset in sequential order, only from 
the top, middle, or bottom, we will likely bias the new dataset and incorrectly train the machine learning algorithm. 
Therefore it is imperative that we engage a random function, guided only by the number of data points for each class.
 
	python karoo_data_sort.py sample.csv

As this sample contains only 20 total rows, select just 5 as the output and it will produce a sorted .csv with 5 of 
each class, 50% of the original file.

The original dataset is left unaltered.	
'''

### USER INTERACTION ###

if len(sys.argv) == 1: print '\n\t\033[31mERROR! You have not assigned an input file. Try again ...\033[0;0m'; sys.exit()
elif len(sys.argv) > 2: print '\n\t\033[31mERROR! You have assigned too many command line arguments. Try again ...\033[0;0m'; sys.exit()
else: filename = sys.argv[1]

os.system('clear')
		
print '\n\033[36m\033[1m'
print '\t **   **   ******    *****    ******    ******       ******    ******    *****   ******'
print '\t **  **   **    **  **   **  **    **  **    **     **    **  **    **  **   **    **'
print '\t ** **    **    **  **   **  **    **  **    **     **        **    **  **   **    **'
print '\t ****     ********  ******   **    **  **    **      ******   **    **  ******     **'
print '\t ** **    **    **  ** **    **    **  **    **           **  **    **  ** **      **'
print '\t **  **   **    **  **  **   **    **  **    **     **    **  **    **  **  **     **'
print '\t **   **  **    **  **   **  **    **  **    **     **    **  **    **  **   **    **'
print '\t **    ** **    **  **    **  ******    ******       ******    ******   **    **   **'
print '\033[0;0m'
print '\t\033[36m Data Prep for Machine Learning in Python - by Kai Staats\033[0;0m\n'

# the UI usually goes here, but in this case we need to know the number of rows in the dataset

### LOAD THE DATA ###

print '\t Loading dataset:\033[36m', filename, '\033[0;0m\n'
data = np.loadtxt(filename, delimiter = ',', dtype = str)
header = data[0]
data = np.delete(data,0,0)
labels = len(np.unique(data[:,-1]))

menu = range(1,len(data)/2)
while True:
	try:
		samples = raw_input('\t Enter number of desired datapoints per class (default %s): ' % str(len(data)/2)) 
		if samples not in str(menu) or samples == '0': raise ValueError()
		elif samples == '': samples = len(data)/2
		samples = int(samples); break
	except ValueError: print '\n\t\033[32m Enter a number from 1 including %s. Try again ... \033[0;0m' % str(len(data)/2)


### SORT DATA BY LABEL ###

data_out = np.empty((1, data.shape[1])) # build an empty array with the same number of columns as the original data

for label in range(labels):
	rows_list = np.where(data[:,-1] == str(label)) # list all rows which end in the current label	
	rows_select = np.random.choice(rows_list[0], samples, replace = False) # randomly select rows from list
	data_out = np.append(data_out, data[rows_select], axis = 0)

data_out = np.delete(data_out,0,0) # delete the top, empty row


### SAVE THE MODIFIED DATA ###

data_out = np.vstack((header, data_out)) # re-attach the header to the data
file_tmp = filename.split('.')[0]
np.savetxt(file_tmp + '-SORT.csv', data_out, delimiter = ',', fmt='%s')

print '\n\t The sorted dataset has been written to the file:\033[36m', file_tmp + '-SORT.csv', '\033[0;0m'




karoo_npy_stack.py
# Karoo Numpy Stack
# by Kai Staats, MSc and Arun Kumar, PhD
# see LICENSE.md
# version 2018 03/26

import os
import sys
import numpy as np 
from scipy.misc import imread

'''
This script prepares a Numpy binary from a set of directory of images. This is terribly useful for processing of
images by Machine Learning algorithms such as a Convolutional Neural Network (CNN) in that the speed at which the 
Machine Learing algorthm can process a Numpy binary is far greater than processing the images individually, and it 
makes them that much more portable.

The current implementation of this script is designed to prepare test images for a CNN conducting Semantic Segementation
on b/w images. It currently works only with the R channel of RGB images, so if your images are in color, two of those 
channels will be lost. A near future version of this script will certainly be more flexible.

	python karoo_npy_stack.py sample_npy_stack/ sample.npy

The original images are left unaltered.	
'''

### USER INTERACTION ###

if len(sys.argv) < 3: print '\n\t\033[31mERROR!\033[0;0m You have assigned too few arguments.\n\n\t[path_to]/[images] [output_image_name].npy. Try again ...\n'; sys.exit()
elif len(sys.argv) > 3: print '\n\t\033[31mERROR!\033[0;0m You have assigned too many arguments.\n\n\t[path_to]/[images] [output_image_name].npy. Try again ...\n'; sys.exit()

os.system('clear')
		
print '\n\033[36m\033[1m'
print '\t **   **   ******    *****    ******    ******      **    **   ******  **    **'
print '\t **  **   **    **  **   **  **    **  **    **     ***   **  **    **  **  **'
print '\t ** **    **    **  **   **  **    **  **    **     ****  **  **    **   ****'
print '\t ****     ********  ******   **    **  **    **     ** ** **  *******     **'
print '\t ** **    **    **  ** **    **    **  **    **     **  ****  **          **'
print '\t **  **   **    **  **  **   **    **  **    **     **   ***  **          **'
print '\t **   **  **    **  **   **  **    **  **    **     **    **  **          **'
print '\t **    ** **    **  **    **  ******    ******      **    **  **          **'
print '\033[0;0m'
print '\t\033[36m Save a stack of images as a Numpy binary - by Arun Aniyan and Kai Staats\033[0;0m'
print ''

### LOAD THE DATA ###

dirname = sys.argv[1] # directory with images
file_out = sys.argv[2] # output filename

dir_list = os.listdir(dirname)
dir_list.sort()

image_x = len(imread(dirname + dir_list[0])) # width of image
image_y = len(imread(dirname + dir_list[1])) # height of image
img_array = np.empty([int(len(dir_list)), int(image_x), int(image_y)])


### GENERATE THE IMAGE ARRAY

i = 0
print '\tLoading images (arrays) ...'

for file_in in dir_list:
	img = imread(str(dirname) + '/' + file_in)
	if len(img.shape) == 3 and img.shape[2] == 3: img_array[i,:,:] = img[:,:,0] # grab the last of the 3-channel (RGB) images
	elif len(img.shape) == 3 and img.shape[2] == 2: img_array[i,:] = img[:,0] # grab the last of the 2-channel (some PNGs) images
	else: img_array[i] = img # for single channel (index'd BMP) images
	i = i + 1
	
if img_array.max() > 1: # the image (array) contains values other than 0 or 1
	print '\n\tOne or more images (arrays) has values other than 1 (black) or 0 (white).'
	
	while True:
		try:
			convert = raw_input('\n\tDo you want to convert values of 255 to -1 and 0 to 1? (y or n): ')
			if convert not in ('y','n',''): raise ValueError()
			if convert == 'n': break
			elif convert == 'y':
				img_array[img_array == 0] = 1; img_array[img_array == 255] = -1
				print '\n\t\033[36mValues of 255 are converted to -1, and values of 0 to 1\033[0;0m'; break
				
		except ValueError: print '\n\t\033[32m Select from the options given, (y)es or (n)o. Try again ...\n\033[0;0m'
		except KeyboardInterrupt: sys.exit()
		
# else: # we are working with a binary (0 or 1) image (array)


### SAVE THE MODIFIED DATA ###

print '\n\tSaving Numpy array file:', file_out

np.save(file_out, img_array)




karoo_multiclassifier.py
# Karoo Multiclass Classifer Test
# by Kai Staats, MSc
# see LICENSE.md
# version 2018 03/26

import os
import numpy as np
from numpy import arange

np.set_printoptions(linewidth = 320) # set the terminal to print 320 characters before line-wrapping in order to view Trees

'''
This is a toy script, designed to allow you to play with multiclass classification using the same underlying function
as employed by Karoo GP. Keep in mind that a linear multiclass classifier such as this is suited only for data which
itself has a linear (eg: time series) component, else GP will struggle to force the data to fit.

	python karoo_multiclassifier
	
'''

### USER INTERACTION ###

os.system('clear')

print '\n\033[36m\033[1m'
print '\t **   **   ******    *****    ******    ******       ******   **      ******    ******    ****** '
print '\t **  **   **    **  **   **  **    **  **    **     **    **  **     **    **  **    **  **    **'
print '\t ** **    **    **  **   **  **    **  **    **     **    **  **     **    **  **        **'
print '\t ****     ********  ******   **    **  **    **     **        **     ********   ******    ******'
print '\t ** **    **    **  ** **    **    **  **    **     **        **     **    **        **        **'
print '\t **  **   **    **  **  **   **    **  **    **     **    **  **     **    **  **    **  **    **'
print '\t **   **  **    **  **   **  **    **  **    **     **    **  **     **    **  **    **  **    **'
print '\t **    ** **    **  **    **  ******    ******       ******    ***** **    **   ******    ******'
print '\033[0;0m'
print '\t\033[36m A multi-classifier demonstration - by Kai Staats\033[0;0m\n'

while True:
	try:
		skew_sel = raw_input('\t Skew the data across the origin? y/n (default y): ')
		if skew_sel not in ['y','n','']: raise ValueError()
		skew_sel = skew_sel or 'y'; break
	except ValueError: print '\n\t\033[32m Select from the options given. Try again ...\033[0;0m'

while True:
	try:
		class_type = raw_input('\t Select (i)nfinite or (f)inite wing bins (default i): ')
		if class_type not in ['i','f','']: raise ValueError()
		class_type = class_type or 'i'; break
	except ValueError: print '\n\t\033[32m Select from the options given. Try again ...\033[0;0m'

menu = range(1,101)
while True:
	try:
		class_labels = raw_input('\t Enter the number of class labels / solutions (default 4): ')
		if class_labels not in str(menu) and class_labels not in '': raise ValueError()
		if class_labels == '0': class_labels = 1; break
		class_labels = class_labels or 4; class_labels = int(class_labels); break
	except ValueError: print '\n\t\033[32m Enter a number from 3 including 100. Try again ...\033[0;0m'


### PROCESS AND OUTPUT TO SCREEN - WITH SKEW ###

if skew_sel == 'y':
	
	skew = (class_labels / 2) - 1
	min_val = 0 - skew - 1 # add a data point to the left
	
	if class_labels & 1: max_val = 0 + skew + 3 # add a data point to the right if odd number of class labels
	else: max_val = 0 + skew + 2 # add a data point to the right if even number of class labels
	
	print '\n\t solutions =', range(class_labels)
	print '\t results = [', min_val, '...', max_val,']'
	print '\t skew =', skew, '\n'
	
	if class_type == 'i':
		for result in arange(min_val, max_val, 0.5):
			for solution in range(class_labels):
			
				if solution == 0 and result <= 0 - skew: # check for the first class
					fitness = 1; print '\t\033[36m\033[1m class', solution, '\033[0;0m\033[36mas\033[1m', result, '\033[0;0m\033[36m<=', 0 - skew, '\033[0;0m'
					
				elif solution == class_labels - 1 and result > solution - 1 - skew: # check for the last class
					fitness = 1; print '\t\033[36m\033[1m class', solution, '\033[0;0m\033[36mas\033[1m', result, '\033[0;0m\033[36m>', solution - 1 - skew, '\033[0;0m'
					
				elif solution - 1 - skew < result <= solution - skew: # check for class bins between first and last
					fitness = 1; print '\t\033[36m\033[1m class', solution, '\033[0;0m\033[36mas', solution - 1 - skew, '<\033[1m', result, '\033[0;0m\033[36m<=', solution - skew, '\033[0;0m'
					
				else: fitness = 0 #; print '\t\033[36m no match for', result, 'in class', solution, '\033[0;0m' # no class match
	
	if class_type == 'f':
		for result in arange(min_val, max_val, .5):
			for solution in range(class_labels):
			
				if solution - 1 - skew < result <= solution - skew: # check for discrete, finite class bins
					fitness = 1; print '\t\033[36m\033[1m class', solution, '\033[0;0m\033[36mas', solution - 1 - skew, '<\033[1m', result, '\033[0;0m\033[36m<=', solution - skew, '\033[0;0m'
					
				else: fitness = 0 #; print '\t\033[36m no match for', result, 'in class', solution, '\033[0;0m' # no class match
				

### PROCESS AND OUTPUT TO SCREEN - WITHOUT SKEW ###

else:

	min_val = 0
	max_val = class_labels
	
	print '\n\t solutions =', range(class_labels)
	print '\t results = [', min_val, '...', max_val,']\n'
	
	if class_type == 'i':
		for result in arange(min_val - 1, max_val, 0.5):
			for solution in range(class_labels):
			
				if solution == 0 and result <= 0: # check for the first class
					fitness = 1; print '\t\033[36m\033[1m class', solution, '\033[0;0m\033[36mas\033[1m', result, '\033[0;0m\033[36m<=', 0, '\033[0;0m'
					
				elif solution == class_labels - 1 and result > solution - 1: # check for the last class
					fitness = 1; print '\t\033[36m\033[1m class', solution, '\033[0;0m\033[36mas\033[1m', result, '\033[0;0m\033[36m>', solution - 1, '\033[0;0m'
					
				elif solution - 1 < result <= solution: # check for classes between first and last
					fitness = 1; print '\t\033[36m\033[1m class', solution, '\033[0;0m\033[36mas', solution - 1, '<\033[1m', result, '\033[0;0m\033[36m<=', solution, '\033[0;0m'
					
				else: fitness = 0 #; print '\t\033[36m no match for', result, 'in class', solution, '\033[0;0m' # no class match
	
	if class_type == 'f':
		for result in arange(min_val, max_val, .5):
			for solution in range(class_labels):
			
				if solution - 1 < result <= solution: # check for discrete, finite class bins
					fitness = 1; print '\t\033[36m\033[1m class', solution, '\033[0;0m\033[36mas', solution - 1, '<\033[1m', result, '\033[0;0m\033[36m<=', solution, '\033[0;0m'
					
				else: fitness = 0 #; print '\t\033[36m no match for', result, 'in class', solution, '\033[0;0m' # no class match




karoo_data_clean.py
# Karoo Data Clean
# by Kai Staats, MSc
# see LICENSE.md
# version 2018 03/26

import os
import sys
import numpy as np

np.set_printoptions(linewidth = 320) # set the terminal to print 320 characters before line-wrapping in order to view Trees

'''
This script loads a .csv dataset which is assumed to have a header and right-most labels (solutions) column, both of 
which are preserved. 

When cleaning by Column, you provide the value you seek to remove from the dataset, and the threshhold (quantity) which 
should invoke the removal. A '0' threshold removes an entire column if even just one instance of the value is found 
while a '100' threshold removes the column only if every single row contains the given value. When cleaning by Row, 
you provide the exact value you seek to remove, and any row that contains this value is automically removed from the 
output dataset.

	python karoo_data_clean.py sample.csv

The original dataset is left unaltered.	
'''

### USER INTERACTION ###

if len(sys.argv) == 1: print '\n\t\033[31mERROR! You have not assigned an external data file. Try again ...\033[0;0m'; sys.exit()
elif len(sys.argv) > 2: print '\n\t\033[31mERROR! You have assigned too many command line arguments. Try again ...\033[0;0m'; sys.exit()
else: filename = sys.argv[1] # you have loaded an external data file

os.system('clear')
		
print '\n\033[36m\033[1m'
print '\t **   **   ******    *****    ******    ******       ******   **      ******   ******   **    **'
print '\t **  **   **    **  **   **  **    **  **    **     **    **  **     **       **    **  ***   **'
print '\t ** **    **    **  **   **  **    **  **    **     **    **  **     **       **    **  ****  **'
print '\t ****     ********  ******   **    **  **    **     **        **     *****    ********  ** ** **'
print '\t ** **    **    **  ** **    **    **  **    **     **        **     **       **    **  **  ****'
print '\t **  **   **    **  **  **   **    **  **    **     **    **  **     **       **    **  **   ***'
print '\t **   **  **    **  **   **  **    **  **    **     **    **  **     **       **    **  **    **'
print '\t **    ** **    **  **    **  ******    ******       ******    *****  ******  **    **  **    **'
print '\033[0;0m'
print '\t\033[36m Data Prep for Machine Learning in Python - by Kai Staats\033[0;0m\n'

while True:
	try:
		clean = raw_input('\t Clean your data by (c)olumn or by (r)ow (default r): ')
		if clean not in ['c','r','']: raise ValueError()
		clean = clean or 'r'; break
	except ValueError: print '\n\t\033[32m Select from the options given. Try again ...\033[0;0m'
	except KeyboardInterrupt: sys.exit()
	
value = raw_input('\t Enter the value you seek in the data (default nan): '); value = str(value) or 'nan'

while True:
	try:
		d_type = raw_input('\t Are you seeking a (s)tring or a (f)loat? (default s): ')
		if d_type not in ['s','f','']: raise ValueError()
		d_type = d_type or 's'; break
	except ValueError: print '\n\t\033[32m Select from the options given. Try again ...\033[0;0m'
	except KeyboardInterrupt: sys.exit()

del_list = []


### LOAD THE DATA ###

print '\n\t Loading dataset:\033[36m', filename, '\033[0;0m'
data = np.loadtxt(filename, delimiter = ',', dtype = str)
header = data[0]
data = np.delete(data,0,0)

if d_type == 'f': data = data.astype(float); value = float(value)

cols = data.shape[1] - 1 # count columns, but exclude the right-most labels column so as to not delete labels
rows = data.shape[0] # count rows
print '\n\t This dataset has\033[36m', cols, '\033[0;0mcolumns (not including labels) and\033[36m', rows, '\033[0;0mrows.'


### COUNT MATCHES ###

if clean == 'r': # clean rows

	for n in range(rows):
		for m in range(cols):
			if data[n][m] == value: del_list.append(n)
				
	if del_list == []: print '\t No rows contain the value:\033[36m', value, '\033[0;0m'; sys.exit()
	else: # at least one row contains the given value
		print '\t The following\033[36m', len(del_list), '\033[0;0mrows will be removed from the new dataset:\n\n\t\033[36m', del_list, '\033[0;0m'
		pause = raw_input('\n\t Press ENTER to continue ...')
		data_clean = np.delete(data, del_list, axis = 0) # remove the designated rows
		header_clean = header
		

elif clean == 'c': # clean columns
	o = np.zeros(cols) # initiate a 1D array with the same x dimension as the original data
	
	for n in range(cols):
		for m in range(rows):
			if data[m][n] == value: o[n] = o[n] + 1
			
		# print '\t column\033[36m', n, '\033[0;0mcontains\033[36m', int(o[n]), '\033[0;0mof the value:\033[36m', value, '\033[0;0m'
		
	if np.sum(o) == 0: print '\t No columns contain the value:\033[36m', value, '\033[0;0m'; sys.exit() # changed .any to .sum 20170929
	else: # at least one column contains the given value
	
		menu = range(0,101)
		while True:
			try:
				threshold = raw_input('\t What is the minimum percental for removal of a column? (default 0): ')
				if threshold not in str(menu): raise ValueError()
				elif threshold == '': threshold = 0
				threshold = int(threshold); break
			except ValueError: print '\t\033[32m Enter a number from 0 including 100. Try again ...\n\033[0;0m'
			except KeyboardInterrupt: sys.exit()
		
		for n in range(cols):
			if o[n] != 0 and float(o[n])/rows >= float(threshold)/100:
				del_list.append(n)

		if del_list == []: print '\n\t No columns contain\033[36m', str(threshold) + '%', '\033[0;0mof the value:\033[36m', value, '\033[0;0m'; sys.exit()	
		else:
			print '\n\t The following columns are removed from the new dataset:\n\n\t\033[36m', del_list, '\033[0;0m'
			pause = raw_input('\n\t Press ENTER to continue ...')
			data_clean = np.delete(data, del_list, axis = 1)
			header_clean = np.delete(header, del_list, axis = 0)
			

### SAVE THE MODIFIED DATA ###

data_out = np.vstack((header_clean, data_clean)) # re-attach the header to the data
file_tmp = filename.split('.')[0]
np.savetxt(file_tmp + '-CLEAN.csv', data_out, delimiter = ',', fmt='%s')

print '\n\t The cleaned dataset has been written to the file:\033[36m', file_tmp + '-CLEAN.csv', '\033[0;0m'


